# Домашнє завдання 08. Random Forest, XGBoost, класифікація і вступ до нейронних мереж

Дата заняття: 18.06.2026 (чт)

Модуль: Модуль 3. Алгоритми навчання з учителем


## Мета
Відпрацювати ансамблі та базові метрики класифікації на зрозумілих supervised learning задачах.

### Завдання
    Створити notebook homework_supervised_models.ipynb.

#### Класифікація цін на житло

##### Завдання

1. Використовуйте California Housing.
2. Перетворіть цільову змінну на бінарну: клас 1, якщо ціна вища за медіану, клас 0, якщо ціна нижча або дорівнює медіані.
3. Для цього обчисліть np.median(y) і створіть y_class = (y > median_price).astype(int).
4. Розділіть дані на train/test (80/20, random_state=42).
5. Навчіть модель LogisticRegression(max_iter=1000).
6. Побудуйте confusion matrix і виведіть Accuracy, Precision, Recall, F1.
7. Напишіть короткий висновок: що показують метрики і яку помилку модель робить частіше.

###### Що здати
    Файл homework_supervised_models.ipynb.


* MedInc  -    Медіанний дохід
* HouseAge - 	Медіанний вік будинку
* AveRooms - 	Середня кількість кімнат
* AveBedrms - 	Середня кількість спалень
* Population -	Населення
* AveOccup - 	Середня кількість мешканців
* Latitude - 	Широта
* Longitude - 	Довгота
* MedHouseVal - 	Медіанна вартість житла

In [37]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, root_mean_squared_error, r2_score

In [38]:
california_housing_df = pd.read_csv('https://raw.githubusercontent.com/dmitriy-shtefan/python-ai/main/data/california_housing.csv')
california_housing_df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847


In [39]:
california_housing_df.describe()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704,2.068558
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532,1.153956
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000,0.149990
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000,1.196000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000,1.797000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000,2.647250
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000,5.000010


In [40]:
california_housing_df.describe()['MedHouseVal']

count    20640.000000
mean         2.068558
std          1.153956
min          0.149990
25%          1.196000
50%          1.797000
75%          2.647250
max          5.000010
Name: MedHouseVal, dtype: float64

In [41]:
X = california_housing_df.drop('MedHouseVal', axis=1)
y = california_housing_df['MedHouseVal']

print(f"X = {X.shape}, y = {y.shape}")

X = (20640, 8), y = (20640,)


In [42]:
median_price = np.median(y)
y_class = (y > median_price).astype(int)

In [43]:
X_train, X_test, y_train, y_test = train_test_split(X, y_class, test_size=0.2, random_state=42) 


In [44]:
from sklearn.linear_model import LogisticRegression


model = LogisticRegression(max_iter=1000, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

for i, col in enumerate(X.columns):
    print(f"{col}: {model.coef_[0][i]}")
print()
print(f"intercept = {model.intercept_}\n"
      f"кількість ітерацій = {model.n_iter_[0]}")


MedInc: 1.559606998833478
HouseAge: 0.03974970174202929
AveRooms: -0.6563797668670918
AveBedrms: 2.9090757505361284
Population: 6.774281002347622e-05
AveOccup: -0.30938531253061013
Latitude: -0.12067751120238832
Longitude: 0.005398581961647954

intercept = [-0.80814081]
кількість ітерацій = 1000


c:\Python\Lib\site-packages\sklearn\linear_model\_logistic.py:599: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [45]:
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score, precision_score, recall_score


cm = confusion_matrix(y_test, y_pred)
print(cm)

accurancy = accuracy_score(y_test, y_pred)
print(f"Частка правильних відповідей серед усіх - {accurancy}") # (TR + TN) / всього   => коли класи збалансовані 0.96

# Зі всіх, кого назвали хворими — скільки справді хворі TP / (TP + FP)
# Коли FP = дорога помилка (спам-фільтр)    => 0.95
precision = precision_score(y_test, y_pred)
print(precision)


# Зі всіх хворих — скільки ми знайшли
# Коли FN = небезпечна помилка (медицина!) => 0.97
recall = recall_score(y_test, y_pred)
print(recall)

# Гармонійне середнє Precision і Recall
# Коли потрібен баланс між P і R  => 0.96

f1 = f1_score(y_test, y_pred)
print(f1)


[[1690  387]
 [ 435 1616]]
Частка правильних відповідей серед усіх - 0.8008720930232558
0.8067898152770844
0.787908337396392
0.7972372964972866


[[1690  387]
 [ 435 1616]]  
True Negative   1690
False Positive  387
False Negative  435
True Positive   1616

Помилки:  - False Positive  387 - ця менш критична, тобто модель біль дешеві будинки визначила як дорогі
        - False Negative  435  -   ця помилка критична, так як дрого будинки визначила як дешеві і втрати під час продажу/купівлі будуть величезними. Ц помилка виникає частіше. 


 - accurancy  приблизно 80% 
 - precision приблизно 80 % 
 - recall приблизно 79% , тобто 21%  дорогих будинків пропущено
 - f1 - баланс витриманий на 79%, але наявні критичні помилки False Negative  435. 

Модель має хорошу загальну точність, але через критичну помилку FN - 435 вона ризикована, для реального використання без додаткового налаштування.